core/vm/contracts.go

```diff
@@ -1353,3 +1358,91 @@ func (c *p256Verify) Run(input []byte) ([]byte, error) {
 		return nil, nil
 	}
 }
+
+// gpgEd25519Verify implements native verification for ed25519 signatures produced via gpg
+type gpgEd25519Verify struct{}
+
+var (
+	errMessageTooShort    = errors.New("message too short")
+	errPubKeyTooShort     = errors.New("public key too short")
+	errSignatureTooShort  = errors.New("signature too short")
+	errInvalidPublicKey   = errors.New("invalid public key format")
+	errInvalidSignature   = errors.New("invalid signature format")
+	errVerificationFailed = errors.New("signature verification failed")
+)
+
+// RequiredGas returns the gas required to execute the pre-compiled contract
+func (c *gpgEd25519Verify) RequiredGas(input []byte) uint64 {
+	// You can adjust this value based on your needs
+	return params.GpgEd25519VerifyGas
+}
+
+// Run performs ed25519 signature verification
+func (c *gpgEd25519Verify) Run(input []byte) ([]byte, error) {
+	// Input should be: message_len (32 bytes) || message || pubkey_len (32 bytes) || pubkey || sig_len (32 bytes) || signature
+	if len(input) < 96 { // minimum length for the three length fields
+		return nil, errMessageTooShort
+	}
+
+	// Extract message length and message
+	msgLen := new(big.Int).SetBytes(input[:32]).Uint64()
+	if len(input) < 32+int(msgLen) {
+		return nil, errMessageTooShort
+	}
+	message := input[32 : 32+msgLen]
+
+	// Extract public key length and public key
+	offset := 32 + msgLen
+	pubKeyLen := new(big.Int).SetBytes(input[offset : offset+32]).Uint64()
+	if len(input) < int(offset+32+pubKeyLen) {
+		return nil, errPubKeyTooShort
+	}
+	pubKey := input[offset+32 : offset+32+pubKeyLen]
+
+	// Extract signature length and signature
+	offset = offset + 32 + pubKeyLen
+	sigLen := new(big.Int).SetBytes(input[offset : offset+32]).Uint64()
+	if len(input) < int(offset+32+sigLen) {
+		return nil, errSignatureTooShort
+	}
+	signature := input[offset+32 : offset+32+sigLen]
+
+	// Convert raw bytes to armored format
+	armoredPubKey := string(pubKey)
+	armoredSig := string(signature)
+
+	// Create public key object
+	pubKeyObj, err := pgpcrypto.NewKeyFromArmored(armoredPubKey)
+	if err != nil {
+		return nil, errInvalidPublicKey
+	}
+
+	// Create public keyring
+	pubKeyRing, err := pgpcrypto.NewKeyRing(pubKeyObj)
+	if err != nil {
+		return nil, errInvalidPublicKey
+	}
+
+	// Parse the armored signature
+	signatureObj, err := pgpcrypto.NewPGPSignatureFromArmored(armoredSig)
+	if err != nil {
+		return nil, errInvalidSignature
+	}
+
+	// Create message object
+	messageHex := hex.EncodeToString(message)
+	messageBytes, err := hex.DecodeString(messageHex)
+	if err != nil {
+		return nil, err
+	}
+	messageObj := pgpcrypto.NewPlainMessage(messageBytes)
+
+	// Verify signature
+	err = pubKeyRing.VerifyDetached(messageObj, signatureObj, pgpcrypto.GetUnixTime())
```

Comment:
They'll have created the signature off chain, and this will only verify if the timestamps match, so `time` should be an input variable (or they should pass a signature that holds that data, can't remember if that exists).

Reply:
You're right. We need to pass a timestamp instead of plainly use `pgpcrypto.GetUnixTime()`. The armoured signature generated from:

```bash
echo [message to sign] | xxd -r -p | gpg --pinentry-mode loopback --detach-sign --armor
```

does not itself contain metadata about the signature creation time. One approach to pass a `verifyTime` may be to enforce a (known, fixed byte length) timestamp to be appended before/after the original message to sign, and have a user sign this original_message+timestamp. That said, I'm not sure if this method is compatible with what tea.xyz is planning to do with GPG signatures.

I checked the source:

```go
// verifySignature verifies if a signature is valid with the entity list.
func verifySignature(
	pubKeyEntries openpgp.EntityList,
	origText io.Reader,
	signature []byte,
	verifyTime int64,
	verificationContext *VerificationContext,
) (*packet.Signature, error) {
	config := &packet.Config{}
	if verifyTime == 0 {
		config.Time = func() time.Time {
			return time.Unix(0, 0)
		}
	} else {
		config.Time = func() time.Time {
			return time.Unix(verifyTime+internal.CreationTimeOffset, 0) // NOTICE!
		}
	}

// ...
}

// CreationTimeOffset stores the amount of seconds that a signature may be
// created in the future, to compensate for clock skew.
const CreationTimeOffset = int64(60 * 60 * 24 * 2)
```

By default, the signature verification in gopenpgp is successful up to two days after the passed `verifyTIme`.

Reply:
Ok, after looking into this more, I believe the setup is:
- the public key contain optional expiration info
- the time passed in verification is supposed to be the present
- the verification checks if the present time within 2 days of the expiration

In our case, we actually want to allow expired public keys to sign, because old GPG keys that signed commits that are being airdropped to _should_ be able to claim, even if they are expired.

Passing `0` bypasses the check, so I believe this is what we should do.